### PR TITLE
fix(explore): Filters with custom SQL disappearing

### PR DESCRIPTION
**Pull request by kgabryje**
### SUMMARY
When user added more than 1 filter with custom SQL and saved the chart, all filters with custom SQL except for the first one disappeared. This PR fixes that issue.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Before:

https://user-images.githubusercontent.com/15073128/185192610-40a9c076-7540-4818-a6b2-b89d1b20c84f.mov

After:

https://user-images.githubusercontent.com/15073128/185192803-44626ed7-e1c6-42e6-ab09-41a7b464c44a.mov

### TESTING INSTRUCTIONS
1. Open explore and add a few filters with custom sql
2. Save chart
3. All added filters should still be there

1. Open a mixed chart
2. Add adhoc filters for both queries
3. Save chart
4. Make sure that adhoc filters for both queries are still there

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/explore/controlUtils/getFormDataWithDashboardContext.ts [hunk 1]
```diff
@@ -176,16 +185,28 @@ export const getFormDataWithDashboardContext = (
...
-    ...ensureIsArray(exploreFormData.adhoc_filters),
-    ...ensureIsArray(filterBoxData.adhoc_filters),
-    ...ensureIsArray(nativeFiltersData.adhoc_filters),
-  ]);
+  const adhocFilters = [
+    ...Object.keys(exploreFormData),
+    ...Object.keys(filterBoxData),
+    ...Object.keys(nativeFiltersData),
+  ]
+    .filter(key => key.match(/adhoc_filter.*/))
```
Comment (kgabryje) [inline superset-frontend/src/explore/controlUtils/getFormDataWithDashboardContext.ts:193]: handles the case of fields like `adhoc_filters_b` in mixed chart. Logic similar to `superset/utils/core.py`, line 1121

**Review by michael-s-molina — APPROVED** [review]
LGTM

**Review by geido — APPROVED** [review]
LGTM!

File: superset-frontend/src/explore/controlUtils/getFormDataWithDashboardContext.ts [hunk 2]
```diff
@@ -51,28 +54,34 @@ const simpleFilterToAdhoc = (
...
     existingFilters.some(
-      (existingFilter: SimpleAdhocFilter) =>
-        existingFilter.operator === adhocFilter.operator &&
-        existingFilter.subject === adhocFilter.subject &&
-        ((!('comparator' in existingFilter) &&
-          !('comparator' in adhocFilter)) ||
-          ('comparator' in existingFilter &&
-            'comparator' in adhocFilter &&
-            isEqual(existingFilter.comparator, adhocFilter.comparator))),
+      (existingFilter: AdhocFilter) =>
```
Comment (michael-s-molina) [inline superset-frontend/src/explore/controlUtils/getFormDataWithDashboardContext.ts:63]: Can we split this code into equality functions for each filter type to improve readability?

**Review by codyml — APPROVED** [review]
Code looks good and fix works locally for me!

**Review by jinghua-qa — APPROVED** [review]
LGTM

**Pull request merged**